CONTRIBUTING.md

```diff
@@ -84,6 +84,21 @@ only on `nightly` toolchain. If you can, test both toolchains.
 * Push your commits to GitHub and create a pull request against Cargo's
 `master` branch.
 
+### Running tests on Windows
+
+To ensure all of Cargo's tests pass on Windows, there are a few things to
+verify with your environment. Besides having a normal Rust development
+environment (with CMake installed), check the following:
+
+* `sh` from MSYS should be in your PATH (ahead of any other `sh` installations
```

Comment:
Something about this isn't correct I think because AppVeyor doesn't have this in PATH, yet tests pass on AppVeyor?

Reply:
Yea, I actually ran a test in AppVeyor that printed the environment because I was confused about that too.  They include `C:\Program Files\Git\usr\bin` in PATH.  [Here](https://gist.github.com/ehuss/d597c70cffe60bc481376f1ce80877e4) is the list.

Reply:
Hm, can you gist again the error that comes about from *not* having `sh` in PATH?

Reply:
The error is at the top of #4591.

Reply:
One workaround for this would be to change git2 to only use `sh` if the helper contains shell-like characters.  I'd be happy to try a PR for that, it's just a bigger change.

Reply:
Oh so as I mentioned on the issue, if that's the only error, we can just fix that by changing [this line](https://github.com/rust-lang/cargo/blob/2f779743f4579febf674dbbb6e24bb2e709a7982/tests/build-auth.rs#L122) to

```
attempted to find username/password via [..]
```

Reply:
If you do that, then it the test just hangs forever because the server thread never receives a connection (because it fails to execute the credential helper).

Reply:
Then... the test should be fixed?

What I'm getting at is that this documentation indicates a bug, and I'm wondering if effort can be applied to fixing the bug instead of documenting it?

Reply:
I think the bug is that git2 assumes you have `sh` in your path.  As I mentioned, I'd be happy to make a PR against git2 to remove that requirement if the helper is just a path, which would fix this issue.

Reply:
I think this may be barking up the wrong tree? It may be the case that some part of `git2` assumes that `sh` is in PATH but if this test is testing *specifically* that then it should just have different behavior when `sh` isn't in PATH

Reply:
We could have the test check if executing `sh` succeeds before starting, and if it fails, skip the test.  However, that doesn't seem like a good idea to me.

Reply:
The test can test something else then which doesn't fail when sh isn't available?

Reply:
I don't think there is anything else the test can do.  As it is, credential helpers need `sh` or they just don't work.

Reply:
So... prefix the test with?

```rust
if Command::new("sh").arg("--version").output().is_err() {
    return 
}
```

Reply:
That wouldn't be my preference (I don't like tests that are silently ignored).  If it were to do something like that, I was thinking it would probably need to run `sh -c path/to/script`.  I don't think `--version` wouldn't be completely safe (for example, I don't think BSD `sh` handles that).  Also, the path would need to use forward slashes on windows to not mysteriously fail with Cygwin.  This is a pretty obscure issue, I was just trying to make it a little less frustrating for future Windows developers to get going.  I'm a little concerned adding all this may make it more brittle.

Reply:
Note that we're not looking for a successful execution of `sh`, just if the program exists.